cardano-api/src/Cardano/Api/ProtocolParametersUpdate.hs

```diff
@@ -0,0 +1,674 @@
+      , protocolUpdateCostModels          = mergeMap protocolUpdateCostModels
+      , protocolUpdatePrices              = mergeMap protocolUpdatePrices
+      , protocolUpdateMaxTxExUnits        = merge protocolUpdateMaxTxExUnits
+      , protocolUpdateMaxBlockExUnits     = merge protocolUpdateMaxBlockExUnits
+      , protocolUpdateParamMaxValSize     = merge protocolUpdateParamMaxValSize
+      }
+      where
+        -- prefer the right hand side:
+        merge :: (ProtocolParametersUpdate -> Maybe a) -> Maybe a
+        merge f = f ppu2 `mplus` f ppu1
+
+        -- need to use map merge where:
+        -- keys present in m1 and m2: remove/ignore m1 key
+        -- keys present in m1 and not m2 : remove/ignore m1 key
+        -- keys present in m2 and not m1: keep m2
+        -- I.e prefer m2 -- TODO: left off here. Remember plutus json jared sent you
+        mergeMap :: (ProtocolParametersUpdate -> Map AnyScriptLanguage a)
+                 -> Map AnyScriptLanguage a
+        mergeMap = undefined
+
+instance Monoid ProtocolParametersUpdate where
+    mempty =
+      ProtocolParametersUpdate {
+        protocolUpdateProtocolVersion     = Nothing
+      , protocolUpdateDecentralization    = Nothing
+      , protocolUpdateExtraPraosEntropy   = Nothing
+      , protocolUpdateMaxBlockHeaderSize  = Nothing
+      , protocolUpdateMaxBlockBodySize    = Nothing
+      , protocolUpdateMaxTxSize           = Nothing
+      , protocolUpdateTxFeeFixed          = Nothing
+      , protocolUpdateTxFeePerByte        = Nothing
+      , protocolUpdateMinUTxOValue        = Nothing
+      , protocolUpdateStakeAddressDeposit = Nothing
+      , protocolUpdateStakePoolDeposit    = Nothing
+      , protocolUpdateMinPoolCost         = Nothing
+      , protocolUpdatePoolRetireMaxEpoch  = Nothing
+      , protocolUpdateStakePoolTargetNum  = Nothing
+      , protocolUpdatePoolPledgeInfluence = Nothing
+      , protocolUpdateMonetaryExpansion   = Nothing
+      , protocolUpdateTreasuryCut         = Nothing
+      , protocolUpdateUTxOCostPerByte     = Nothing
+      , protocolUpdateCostModels          = mempty
+      , protocolUpdatePrices              = mempty
+      , protocolUpdateMaxTxExUnits        = Nothing
+      , protocolUpdateMaxBlockExUnits     = Nothing
+      , protocolUpdateParamMaxValSize     = Nothing
+      }
+
+-- ----------------------------------------------------------------------------
+-- Proposals embedded in transactions to update protocol parameters
+--
+
+data UpdateProposal =
+     UpdateProposal
+       !(Map (Hash GenesisKey) ProtocolParametersUpdate)
+       !EpochNo
+    deriving stock (Eq, Show)
+
+instance HasTypeProxy UpdateProposal where
+    data AsType UpdateProposal = AsUpdateProposal
+    proxyToAsType _ = AsUpdateProposal
+
+instance HasTextEnvelope UpdateProposal where
+    textEnvelopeType _ = "UpdateProposalShelley"
+
+--TODO: Jordan UpdateProposal needs to be parameterized by era or have access to the era
```

Comment:
If we think we might need the era just for this purpose, then we can either:
1. Drop CBOR serialisation for these types. Do we really need separate serialisation for them?
2. Do independent CBOR serialisation that does not involve converting to the underlying ledger type. This will not work in future precisely because we're taking the simple untyped approach of taking the union of all params for all eras.